pkg/server/service.go

```diff
@@ -0,0 +1,104 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package server
+
+import (
+	"encoding/json"
+	"time"
+
+	monitorAPI "github.com/cilium/cilium/pkg/monitor/api"
+	"go.uber.org/zap"
+)
+
+const (
+	serviceCacheInitRetryInterval = 5 * time.Second
+	serviceCacheRefreshInterval   = 5 * time.Minute
+)
+
+// fetchServiceCache fetches the service cache from cilium and initializes the
+// local service cache.
+func (s *ObserverServer) fetchServiceCache() error {
+	entries, err := s.ciliumClient.GetServiceCache()
+	if err != nil {
+		return err
+	}
+	if err := s.serviceCache.InitializeFrom(entries); err != nil {
+		return err
+	}
+	s.log.Debug("Fetched service cache from cilium", zap.Int("entries", len(entries)))
+	return nil
+}
+
+// processServiceEvent decodes and applies a service update. It returns true
+// when successful.
+func (s *ObserverServer) processServiceEvent(an monitorAPI.AgentNotify) bool {
+	switch an.Type {
+	case monitorAPI.AgentNotifyServiceUpserted:
+		n := monitorAPI.ServiceUpsertNotification{}
+		if err := json.Unmarshal([]byte(an.Text), &n); err != nil {
+			s.log.Error("Unable to unmarshal service upsert notification",
+				zap.Int("type", int(an.Type)), zap.String("ServiceUpsertNotification", an.Text))
+			return false
+		}
+		return s.serviceCache.Upsert(int64(n.ID), n.Name, n.Type, n.Namespace, n.Frontend.IP, int(n.Frontend.Port))
+	case monitorAPI.AgentNotifyServiceDeleted:
+		n := monitorAPI.ServiceDeleteNotification{}
+		if err := json.Unmarshal([]byte(an.Text), &n); err != nil {
+			s.log.Error("Unable to unmarshal service delete notification",
+				zap.Int("type", int(an.Type)), zap.String("ServiceDeleteNotification", an.Text))
+			return false
+		}
+		return s.serviceCache.DeleteByID(int64(n.ID))
+	default:
+		s.log.Warn("Received unknown service notification type", zap.Int("type", int(an.Type)))
+		return false
+	}
+}
+
+func (s *ObserverServer) syncServiceCache(serviceEvents <-chan monitorAPI.AgentNotify) {
+	for err := s.fetchServiceCache(); err != nil; err = s.fetchServiceCache() {
+		s.log.Error("Failed to fetch service cache from Cilium", zap.Error(err))
+		time.Sleep(serviceCacheInitRetryInterval)
+	}
+
+	refresh := time.NewTimer(serviceCacheInitRetryInterval)
+	inSync := false
+
+	for serviceEvents != nil {
+		select {
+		case <-refresh.C:
+			if err := s.fetchServiceCache(); err != nil {
+				s.log.Error("Failed to fetch service cache from Cilium", zap.Error(err))
+			}
+			refresh.Reset(serviceCacheInitRetryInterval)
+		case an, ok := <-serviceEvents:
+			if !ok {
+				return
+			}
+			// Initially we might see stale updates that were enqued before we
+			// initialized the ipcache. Once we see the first applicable update
+			// though, all subsequent updates must be applicable as well.
+			updated := s.processServiceEvent(an)
+			switch {
```

Comment:
While I added this "is caught up" logic initially originally in `ipcache.go` myself, I'm honestly not sure if it carries its weight, especially for the service cache which has immutable service IDs (which the IP cache apparently doesn't have). Feel free to just remove it if you feel the same.